rfc/rfc-0005-customizable-consumption-offset.md

```diff
@@ -0,0 +1,66 @@
+# RFC-0005: Optional starting offset for processor inputs
+
+**Authors:** Eric Bottard
+
+**Status:**
+
+**Pull Request URL:** https://github.com/projectriff/riff/pull/1367
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** rfc-0003: developer utils
+
+
+## Problem
+Sometimes users want to replay data through functions, maybe because the function logic has changed, or maybe because downstream systems have lost state and that state needs to be re-created. The current Processor concept in riff doesn't support such a mechanism though, and always consumes data from the (current) end of streams.
+
+### Anti-Goals
+
+## Solution
+With a top to bottom view of the user experience, the proposed solution is to
+- introduce an offset option on each input binding when a processor is created
+- persist that option in the Processor CRD instance on the cluster
+- have the Processor reconciler forward that option to the streaming-processor, so that it changes its SubscriptionRequests accordingly
+
+### User Impact
+This RFC proposes to change the `Inputs` field of the `Processor` CRD from
+```go
+Inputs []StreamBinding `json:"inputs"`
+```
+to
+```go
+Inputs []InputStreamBinding `json:"inputs"`
+...
+type InputStreamBinding struct {
+    StreamBinding
+	// Where to start consuming this stream the first time a processor runs.
+    StartOffset int64 `json:"startOffset"`
+}
+```
+
+with `StartOffset` supporting only two special, negative values for now: earliest and latest (default).
+
+This RFC proposes to change the `riff streaming processor create` command from
+```
+riff streaming processor create <name> ... --input [alias:]stream ...
+```
+to
+```
+riff streaming processor create <name> ... --input [alias:]stream[:offset] ...
+```
+where `[..]` denotes optional values and `[:offset]` could take values `start` or `end`. Those two logical values are the only ones currently supported by liiklus (and the high level kafka consumer as far as the author can remember). Nevertheless, the encoding of those two values could be done via negative numerical values, in preparation of future support of an actual numerical index.
```

Comment:
reading after `end` seems a little confusing. I like `earliest` and `latest`.

Reply:
I agree that "end" is confusing since the stream may "never end". Similarly, "start" is misleading in the case where the stream has been truncated based on the broker's retention policy.

It does seem that "earliest" and "latest" most accurately describe where we want to start consuming based on the currently stored events, and those are the 2 values accepted for the Kafka consumer's `auto.offset.reset` property as listed in the table here: https://kafka.apache.org/documentation/#consumerconfigs